float64[] q_array